core/data_access/base_models/base_model.py

```diff
@@ -0,0 +1,37 @@
+from django.utils import timezone
+from django.db import models
+from core.data_access.base_models.model_event import ModelEvent
+
+
+class BaseModel(models.Model):
+    _events = []
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+   
+    class Meta:
+        abstract = True
+        
+    def add_event(self, event: ModelEvent):
+        """Add an event to the list of events."""
+        self._events.append(event)
+        
+    def _clear_events(self):
+        """Clear the list of events."""
+        self._events = []
+
+    def dispatch_events(self):
+        """Dispatch all events."""
+        if self._events.count == 0: 
+            return
+        for event in self._events:
+            from core.setup.mediator_setup import get_mediator  # Local import to avoid circular dependencies
```

Comment:
The circular import issue arises because:

BaseModel --> from mediator_setup Import get_mediator
mediator_setup --> Import PlayerCreatedEvent
PlayerCreatedEvent --> Import Player
Player --> Import BaseModel

Which is full circle -- but in a bad way.

We can get around this by using something called "TYPE CHECKING" Which means we only do imports for the typechecking

e.g.

from typing import TYPE_CHECKING

from core.data_access.base_models.model_event import ModelEvent

if TYPE_CHECKING:
    from core.data_access.models.player.player_model import Player
    from core.data_access.models.verification_code_model import VerificationCode


class PlayerCreatedEvent(ModelEvent):
    def __init__(self, code: 'VerificationCode', player: 'Player'):
        self.verification_code = code
        self.player = player
        super().__init__()

Reply:
I can set this to resolved but I think it should be a future task to use TYPE_CHECKING when we only import for typehinting